### PR TITLE
ci(actions): run amd64 E2E on GitHub actions and arm64 on CircleCI

### PR DESCRIPTION
**Pull request by jijiechen**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - N/A
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS
  - Confirmed
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - Manually tested
  - Don't forget `ci/` labels to run additional/fewer tests
    - Added
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)?
  - No need
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) 
  - No need to back-port.

This PR follows the previous one at [here](https://github.com/kumahq/kuma/pull/8457) and supports running E2E tests on GitHub Actions when `arch is amd64`, other arches will still run on CircleCI

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: .github/workflows/build-test-distribute.yaml [hunk 1]
```diff
@@ -120,9 +120,14 @@ jobs:
...
     needs: ["check", "test", "test_e2e", "test_e2e_env"]
-    if: ${{ always() && !failure() && !cancelled() }}
+    if: ${{ always() }}
     runs-on: ubuntu-latest
     steps:
+      # see https://stackoverflow.com/a/67532120/4907315
+      - name: "Halt due to previous failures"
+        run: |
+          exit 1
+        if: ${{ contains(needs.*.result, 'failure')|| contains(needs.*.result, 'cancelled') }}
```
Comment (jijiechen) [inline .github/workflows/build-test-distribute.yaml:130]: For some reason, `disttributions` still run even if some of `test_e2e`  failed. So we check manually here.
Here is an example of it: https://github.com/jijiechen/kuma/actions/runs/7003299010/attempts/1

File: .github/workflows/build-test-distribute.yaml [hunk 2]
```diff
@@ -34,19 +34,14 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
+      # GitHub actions does not share cache across multiple jobs,
```
Comment (jijiechen) [inline .github/workflows/build-test-distribute.yaml:37]: shifting these snippets to make it more readable and align ordering with other jobs.

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -29,7 +29,104 @@ on:
...
+      - name: Install build tools
+        run: |
+          command: |
+            if ! command -v sudo 2>&1 >/dev/null; then
+              apt update
+              apt install -y sudo
+            fi
+
+            sudo apt update
+            sudo env DEBIAN_FRONTEND=noninteractive apt install -y curl git make unzip gcc xz-utils
```
Comment (lahabana) [inline .github/workflows/e2e.yaml:72]: Do we need these? I think things like curl, git, make unzip are already installed no

**Review by lahabana — APPROVED** [review]
LGTM!

**Pull request merged**